api-service/webapp-demo/src/app/base_job.component.spec.ts

```diff
@@ -0,0 +1,344 @@
+import { ActivatedRoute, ParamMap, Router } from '@angular/router';
+import {
+  HttpClientTestingModule,
+  HttpTestingController
+} from '@angular/common/http/testing';
+import {
+  BaseRequestOptions,
+  Http,
+  Response,
+  ResponseOptions,
+  XHRBackend
+} from '@angular/http';
+import { By } from '@angular/platform-browser';
+import { BehaviorSubject } from 'rxjs/BehaviorSubject';
+import { BaseJobComponent } from './base_job.component';
+import { CrowdSourceApiService } from './crowd_source_api.service';
+
+// A simple Component that extends BaseJobComponent.
+@Component({
+  selector: 'my-job',
+  template: `
+    <div *ngIf="question !== null">{{questionId}} {{question.text}}</div>
+    <button id="sendScoreButton" (click)="sendScoreToApi()">Send score</button>
+  `
+})
+class MyJobComponent extends BaseJobComponent {
+  customClientJobKey = 'testJobKey';
+  routerPath = '/my_job';
+}
+
+// Test component wrapper for BaseJobComponent.
+@Component({
+  selector: 'base-job-test',
+  template: "<base-job></base-job>"
+})
+class BaseJobTestComponent {
+}
+
+// Test component wrapper for MyJobComponent.
+@Component({
+  selector: 'base-job-extended-test',
+  template: "<my-job></my-job>"
+})
+class BaseJobExtendedTestComponent {
+  @ViewChild(MyJobComponent) myJobComponent: MyJobComponent;
+}
+
+// Stub for ActivatedRoute.
+@Injectable()
+class ActivatedRouteStub {
+  private paramsSubject = new BehaviorSubject(this.testParams);
+  paramMap = this.paramsSubject.asObservable();
+
+  private stubTestParams: ParamMap;
+  get testParams() {
+    return this.stubTestParams;
+  }
+  set testParams(params: {}) {
+    this.stubTestParams = new ParamMapStub(params);
+    this.paramsSubject.next(this.stubTestParams);
+  }
+}
+
+// A stub class that implements ParamMap.
+class ParamMapStub implements ParamMap {
+  constructor(private params: {[key: string]: string}) {}
+  has(name: string): boolean {
+    return this.params.hasOwnProperty(name);
+  }
+  get(name: string): string | null {
+    return this.params[name];
+  }
+
+  getAll(name: string): string[] {
+    return [this.params[name]];
+  }
+
+  get keys(): string[] {
+    let objKeys = [];
+    for(let key in this.params) {
+      objKeys.push(key);
+    }
+    return objKeys;
+  }
+}
+
+// TODO(rachelrosen): Currently we have to expect every http call in order to
+// be able to call .verify() at the end; investigate if there is another way
+// to "flush" the controller so that we don't have to check the job_quality
+// and worker quality calls are made in every test where we want to use
+// verify().
+function verifyQualityApiCalls(httpMock: HttpTestingController) {
+  // Verify job quality call.
+  httpMock.expectOne('/api/job_quality').flush({
+    toanswer_count: 50,
+    toanswer_mean_score: 2
+  });
+
+  // Verify worker quality call.
+  httpMock.expectOne((req: HttpRequest<any>): boolean => {
+    const workerQualityRequestUrlRegExp = /\/api\/quality\/?(\w+)?/;
```

Comment:
Not a big deal, but the `user_nonce` is always going to be a string of a number between 0 and 1 the way it's generated now. For example, `0.3824493160515674`. So you could restrict this regex a bit more to capture that. 

https://github.com/conversationai/conversationai-crowdsource/blob/master/api-service/webapp-demo/src/app/base_job.component.ts#L97